### PR TITLE
add proxy plugin workflow

### PR DESCRIPTION
**Pull request by evgeni**
needs: https://github.com/ekohl/ruby-version/pull/1

### CODE REVIEW

File: .github/workflows/proxy_plugin.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        id: build_matrix
+        uses: theforeman/gha-matrix-builder@v0
+        with:
+          repository: ${{ inputs.foreman_proxy_repository }}
+          ref: ${{ inputs.foreman_proxy_version }}
+          include: ${{ inputs.matrix_include }}
+          exclude: ${{ inputs.matrix_exclude }}
+
+  test:
+    name: "Foreman Proxy ${{ inputs.foreman_proxy_version }} with Ruby ${{ matrix.ruby }}"
```
Comment (evgeni) [inline .github/workflows/proxy_plugin.yml]: this is a lie, we take the matrix from that version, but we never setup that proxy ourself

File: .github/workflows/proxy_plugin.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
+---
+name: Foreman Proxy Plugin
```
Comment (ehelms) [inline .github/workflows/proxy_plugin.yml]: Bikeshedding -- given the code projects are all named smart_proxy why choose Foreman Proxy here?

Comment (evgeni) [reply .github/workflows/proxy_plugin.yml]: I created this file based on `foreman_plugin.yml` and replacing it that way was quicker ;)

(also, you could argue the file itself should be called `foreman_proxy_plugin` or `smart_proxy_plugin`)

At the very end, I don't care either way :)

File: .github/workflows/proxy_plugin.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        required: false
+        type: string
+      matrix_exclude:
+        description: Test matrix combinations to exclude
+        default: '[]'
+        required: false
+        type: string
+
+env:
+  BUNDLE_WITHOUT: "bmc:development:dhcp_isc_inotify:dhcp_isc_kqueue:journald:krb5:libvirt:puppetca_token_whitelisting:realm_freeipa:windows"
```
Comment (ekohl) [inline .github/workflows/proxy_plugin.yml]: I've been thinking about making the gemfiles smarter. Like using `RUBY_PLATFORM` to detect if it's BSD or Linux. Something like https://github.com/theforeman/smart-proxy/pull/884.

And I'm not sure we should be so aggressive in disabling all the modules. Or we can provide an env var that toggles `optional` in the groups.

Comment (evgeni) [reply .github/workflows/proxy_plugin.yml]: I copied that list from Jenkins :woman_shrugging: 

We surely can be smarter, but this was not part of the exercise ;-)

**Review by ekohl — COMMENTED** [review]
I'm wondering if this makes sense at all. Unlike Foreman plugins, Smart Proxy plugins are plain gems that can be installed (by depending on `smart_proxy` in `Gemfile`). So you can just use our existing [`gem-test`](https://github.com/theforeman/actions#gem-test) workflow. It respects `required_ruby_version` in the gemspec and tests on all of them.

You can argue that we somehow want to enforce that it includes all the Ruby versions we care about, but is this the way to go?

**Comment by evgeni** [comment]
Well, I like opinionated solutions.
Forcing the matrix to match the one from the main proxy repo is nice.
Also allowing testing with different proxy versions (even if that's not implemented today, we could do it similarly how we test different puppets by looking at ENV[PUPPET_VERSION])

File: .github/workflows/proxy_plugin.yml (hunk 2 above)
Comment (ehelms) [reply .github/workflows/proxy_plugin.yml]: Fair, `smart_proxy_plugin` would be more accurate and connect more cleanly.

File: .github/workflows/smart_proxy_plugin.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+---
+name: Foreman Proxy Plugin
+
+on:
+  workflow_call:
+    inputs:
+      foreman_proxy_version:
+        description: Foraman proxy version to use
```
Comment (ekohl) [inline .github/workflows/smart_proxy_plugin.yml]:
```suggestion
        description: Foreman Proxy version to use
```

File: .github/workflows/smart_proxy_plugin.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        default: 'develop'
+        required: false
+        type: string
+      foreman_proxy_repository:
+        description: Foreman proxy repository to use
```
Comment (ekohl) [inline .github/workflows/smart_proxy_plugin.yml]:
```suggestion
        description: Foreman Proxy repository to use
```

**Review by ekohl — COMMENTED** [review]
Do we want to add a "Test suite" check at the end that we can mark as a required check?

**Comment by evgeni** [comment]
> Do we want to add a "Test suite" check at the end that we can mark as a required check?

There already is https://github.com/theforeman/actions/blob/0ba3a0fcfb56ae415cf92f9090f5fdae71ba2888/.github/workflows/test-gem.yml#L59-L60
Can't this be used as the "required" one?

**Comment by ekohl** [comment]
Oh yes, I suppose so.

Can we create a workflow somewhere to test this out? Perhaps in this repo itself, providing CI for CI?

**Comment by evgeni** [comment]
Way ahead of you ;)
* https://github.com/theforeman/smart_proxy_pulp/pull/36
* https://github.com/Katello/smart_proxy_container_gateway/pull/41

**Comment by evgeni** [comment]
For "in this repo" we'd need to teach `test-gem` to check out arbitrary repos -- should we?

File: .github/workflows/smart_proxy_plugin.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    name: Verify supported Ruby versions
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Verify supported Ruby versions
+        uses: theforeman/gha-matrix-ruby-verifier@v0
+        with:
+          repository: ${{ inputs.foreman_proxy_repository }}
+          ref: ${{ inputs.foreman_proxy_version }}
+  test:
```
Comment (ekohl) [inline .github/workflows/smart_proxy_plugin.yml:40]: Can we make it easy to respect `inputs.foreman_proxy_version` in `Gemfile`? Like set an env var that we can use. Or is there some other way?

**Review by ekohl — CHANGES_REQUESTED** [review]

**Comment by ekohl** [comment]
> For "in this repo" we'd need to teach `test-gem` to check out arbitrary repos -- should we?

I wouldn't mind. I suppose an alternative is to introduce some dummy gemspec, but not sure if that's any better.

File: .github/workflows/proxy_plugin.yml (hunk 1 above)
Comment (evgeni) [reply .github/workflows/proxy_plugin.yml]: We now do, as we pass `SMART_PROXY_BRANCH` env var, which will carry the right branch to check out in Gemfile of the plugin

File: .github/workflows/smart_proxy_plugin.yml (hunk 6 above)
Comment (evgeni) [reply .github/workflows/smart_proxy_plugin.yml:40]: env var it is

**Comment by evgeni** [comment]
@ekohl implemented self-testing :)

**Comment by evgeni** [comment]
The `[TMP]` commits gotta be dropped before merging :)

File: .github/workflows/test-gem.yml [hunk 7]
```diff
@@ -23,7 +36,10 @@ jobs:
       ruby: ${{ steps.ruby.outputs.versions }}
     steps:
       - id: ruby
-        uses: ekohl/ruby-version@v0
+        uses: evgeni/ruby-version@multipass
```
Comment (ekohl) [inline .github/workflows/test-gem.yml]: I merged your PR
```suggestion
        uses: ekohl/ruby-version@v0
```

File: .github/workflows/smart_proxy_plugin_workflow_test.yml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  cancel-in-progress: true
+
+jobs:
+  test:
+    name: Ruby
+    uses: ./.github/workflows/smart_proxy_plugin.yml
+    with:
+      plugin_repository: theforeman/smart_proxy_pulp
+      plugin_version: gha
+      foreman_proxy_version: enter-the-matrix
```
Comment (ekohl) [inline .github/workflows/smart_proxy_plugin_workflow_test.yml]: I merged this

**Review by ekohl — COMMENTED** [review]
Overall ACK but needs an update now that the required PRs have been merged.

**Comment by evgeni** [comment]
@ekohl this still contains two TMP commits that need dropping before merging, tests are green right now. If you ack, I'll drop the two and merge?

**Review by ekohl — APPROVED** [review]

**Pull request merged**